### PR TITLE
Set up pytest node tests

### PR DESCRIPTION
**Pull request by hoodmane**

### CODE REVIEW

**Comment by hoodmane** [comment]
`globalThis` is weird in Firefox...

**Comment by hoodmane** [comment]
Okay we're up to 666/735 core node tests passing and 7 I've xfailed because they use many browser-only APIs. Of the 68 remaining failures, 60 are cascading failures due to timeouts in `test_core_python`. After a timeout my node driver should probably be refreshed. Or maybe the communication protocol can somehow be fixed up to be able to clean up old garbage in stdout. Might require some work though. Maybe I can raise the default timeout in the short term.

Also, the node tests are REALLY slow. I'm not sure why they are so slow tbh, but it's actually about similar to what happened on my libffi port. There, node tests take 15x longer (sometimes even more), here they only take 6x longer than the selenuim tests.

**Comment by hoodmane** [comment]
`test-core-node: 2 failed, 582 passed, 9 skipped`, and the two failures passed on the retry. So that's 584 tests passing. For comparison there are 595 that run in `test-core-firefox` / `test-core-chrome`, so only 11 node-related xfails in core.

**Comment by hoodmane** [comment]
It does take 6 times longer to run the node test suite than the firefox one though...

**Comment by hoodmane** [comment]
So `test-packages-node` works fine up to `test_numpy` where it hangs.

**Comment by hoodmane** [comment]
Okay so the state of affairs here is: 
* `test-core-node` is all good and working.
* The 13 matplotlib tests are all xfailed, they don't make sense in node.
* The numpy tests hang on CI. Locally they pass fine, more investigation needed. Currently skipping all 56 numpy tests.
* I xfailed `scikit-image` and `statsmodels`. They failed a few times and they are xfailed for chrome. It would make sense if the trouble is with v8.
* PILLOW test timing out a lot, might xfail it if it times out again
* All other package tests seem to work on node

Interestingly with my node test driver, the time difference between `selenium_standalone` and `selenium` doesn't seem very large. `test-core-node` takes 6x as long as `test-core-chrome` but `test-packages-node` only takes 1.25x as long.

File: conftest.py [hunk 1]
```diff
@@ -350,6 +300,85 @@ def get_driver(self):
...
+        os.chdir("..")
+
+    def get_driver(self):
+        self._logs = []
+        self.init_node()
+
+        class NodeDriver:
+            def __getattr__(self, x):
+                raise Exception("BLAHH", x)
+                # pytest.skip("unsupported configuration")
```
Comment (rth) [inline conftest.py:325]: Please remove commend and make the exception message a bit more explicit. Maybe a `NotImplementedError` instead ?

**Review by rth — COMMENTED** [review]
Thanks for working on this!

xfailing those tests sound good to me.

> Interestingly with my node test driver, the time difference between selenium_standalone and selenium doesn't seem very large. test-core-node takes 6x as long as test-core-chrome but test-packages-node only takes 1.25x as long.

Might be due to a 50ms sleep by default after each send command and you are sending line by line.. https://stackoverflow.com/a/60423834/1791279

I think we would need to figure out how to make the `test-core[node]` faster (if the above idea doesn't work) as otherwise >1h run time is going to difficult to manage in CI and it might mean that we won't be able to run it on PRs (since we only have 4 parallel jobs in CircleCI)

File: testsetup.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
+Error.stackTraceLimit = Infinity;
+// Firefox globalThis is screwed up. Replace with window.
```
[the remark this answers is not in the document]
Comment (hoodmane) [reply testsetup.js:2]: See https://github.com/facebook/react/issues/16606

I finally ended up just using `self` everywhere as a workaround, it's easier to polyfill `self` in `node` with `globalThis.self = globalThis.window || globalThis`. So this polyfill is not needed anymore.

**Comment by hoodmane** [comment]
> due to a 50ms sleep by default after each send command and you are sending line by line

Hmm. I guess I can try sending all at once.

> I think we would need to figure out how to make the test-core[node] faster.

I am working on embedding node into a Python extension so we can avoid this `pexpect` nonsense. It is working decently so far (I can `run_js` a Javascript code string and get back the return value). The biggest problem with this approach is that `node` doesn't seem to ship `libnode.so` or the other shared libraries from a build. It took around three hours to build node on my local machine. I am just building node in the most standard configuration though:
```sh
./configure --shared
make -j4
```
so I think it would be okay to stash libnode somewhere if we can't find someone who has already done this.

**Comment by rth** [comment]
> Hmm. I guess I can try sending all at once.

In the above SO link, there was also the option of setting that sleep value to 0 which should help.

> so I think it would be okay to stash libnode.so somewhere if we can't find someone who has already done this.

Yeah, we can't reasonably ask people to build node from sources. I think shipping a `.so` also likely wouldn't work as that's not portable: it won't work for Linux/MacOS users who are trying to build outside of Docker.

Maybe profiling a typical pexpect session to understand what's going on would help.

**Comment by hoodmane** [comment]
> we can't reasonably ask people to build node from sources

Well this isn't a key aspect of the project, I think we could just tell people that to run node tests they should either use docker or acquire a `libnode.so` that works on their system?

**Comment by hoodmane** [comment]
Apparently `electron` uses `libnode.so` so maybe we can find it in electron distros?

**Comment by rth** [comment]
True, but it would overall make contributing more complex if say one of the node tests failed and one need to reproduce. Also even to include that .so into Docker doesn't sound straightforward. I would rather not compile node even for building Docker images.

My point is that it would be good to first try to understand why pexpect is slow. Serializing strings for input/output shouldn't be that much overhead.

**Comment by hoodmane** [comment]
Yeah "embed node into a python app" is a bit in the fun but logistically impractical direction. Also, it's pretty incredible how much better the docs for the Python C api are than for the Node C++ API. The Node / v8 API seems significantly more complicated but the docs are close to nonexistent.

**Comment by rth** [comment]
> Apparently electron uses libnode.so so maybe we can find it in electron distros?

I'm not too familiar with electron distros, but I agree that one of the challenges we have is really to keep the tools we use as simple as possible. It's already and a very complex build setup, so the more additional tools we use the more difficult it is for new contributors to approach it (and also for us to have solid enough expertise in those tools).

**Comment by hoodmane** [comment]
Here's the discussion where people said it would be good to ship an `so` (to avoid forcing embedders to build it):
https://github.com/nodejs/node/issues/24028

There is also [metacall](https://github.com/metacall/core/) but documentation is very thin.

**Comment by hoodmane** [comment]
It looks like electron links `libnode.a` and then doesn't distribute it.

**Comment by hoodmane** [comment]
Yeah hopefully we can make `pexpect` work, turns out embedding node is a huge can of worms.

**Comment by hoodmane** [comment]
Looks like `delaybeforesend=None` did the trick, now node tests run the fastest of the three.

File: testsetup.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (hoodmane) [reply testsetup.js]: Never mind I guess it actually is needed.

**Comment by hoodmane** [comment]
Looks like we already xfailed the pillow test on chrome and firefox. I think this is all ready now.

**Comment by hoodmane** [comment]
@rth Tests all pass. I think this is ready to merge, unless you have further comments.

**Comment by rth** [comment]
Thanks for working on this @hoodmane !

**Pull request merged**